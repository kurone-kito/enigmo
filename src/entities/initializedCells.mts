import { CELLS, COLORS, SP_POS, WALLS } from '../constants.mjs';
import type {
  Cell,
  Cells,
  CellsIndex,
  ColorsIndex,
  Special,
  SpecialsIndex,
  WallsIndex,
} from './types.mjs';

/** The offset of the special cells. */
const specials = Object.freeze(
  [1, 0, 2].map<Special>((item, index) => {
    const sIndex = item as SpecialsIndex;
    const initial = (SP_POS + index) as CellsIndex;
    return Object.freeze<Special>({ initial, sIndex, type: 'special' });
  }),
);

/**
 * Initialize the cells.
 * @returns the cells
 */
const initializeCells = (): Cells =>
  COLORS.reduce<readonly Cell[]>(
    (cAcc, _, ci) =>
      WALLS.reduce<readonly Cell[]>((acc, __, wi) => {
        const cIndex = ci as ColorsIndex;
        const wIndex = wi as WallsIndex;
        const length = acc.length;
        const ins = length - SP_POS === 0;
        const initial = (length + (ins ? specials.length : 0)) as CellsIndex;
        const wall = { cIndex, initial, wIndex, type: 'wall' } as const;
        return length >= CELLS ? acc : [...acc, ...(ins ? specials : []), wall];
      }, cAcc),
    [],
  ) as Cells;

/** The initialized cells. */
export const initializedCells = initializeCells();
