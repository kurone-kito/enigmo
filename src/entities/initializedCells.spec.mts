import { describe, expect, it } from 'vitest';
import { initializedCells } from './initializedCells.mjs';

describe('initializedCells', () => {
  it('should be initialized', () => expect(initializedCells).toMatchSnapshot());
});
