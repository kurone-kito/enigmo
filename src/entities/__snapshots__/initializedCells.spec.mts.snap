// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`initializedCells > should be initialized 1`] = `
[
  {
    "cIndex": 0,
    "initial": 0,
    "type": "wall",
    "wIndex": 0,
  },
  {
    "cIndex": 0,
    "initial": 1,
    "type": "wall",
    "wIndex": 1,
  },
  {
    "cIndex": 0,
    "initial": 2,
    "type": "wall",
    "wIndex": 2,
  },
  {
    "cIndex": 0,
    "initial": 3,
    "type": "wall",
    "wIndex": 3,
  },
  {
    "cIndex": 0,
    "initial": 4,
    "type": "wall",
    "wIndex": 4,
  },
  {
    "cIndex": 0,
    "initial": 5,
    "type": "wall",
    "wIndex": 5,
  },
  {
    "cIndex": 0,
    "initial": 6,
    "type": "wall",
    "wIndex": 6,
  },
  {
    "cIndex": 0,
    "initial": 7,
    "type": "wall",
    "wIndex": 7,
  },
  {
    "cIndex": 1,
    "initial": 8,
    "type": "wall",
    "wIndex": 0,
  },
  {
    "cIndex": 1,
    "initial": 9,
    "type": "wall",
    "wIndex": 1,
  },
  {
    "cIndex": 1,
    "initial": 10,
    "type": "wall",
    "wIndex": 2,
  },
  {
    "cIndex": 1,
    "initial": 11,
    "type": "wall",
    "wIndex": 3,
  },
  {
    "cIndex": 1,
    "initial": 12,
    "type": "wall",
    "wIndex": 4,
  },
  {
    "cIndex": 1,
    "initial": 13,
    "type": "wall",
    "wIndex": 5,
  },
  {
    "cIndex": 1,
    "initial": 14,
    "type": "wall",
    "wIndex": 6,
  },
  {
    "cIndex": 1,
    "initial": 15,
    "type": "wall",
    "wIndex": 7,
  },
  {
    "cIndex": 2,
    "initial": 16,
    "type": "wall",
    "wIndex": 0,
  },
  {
    "cIndex": 2,
    "initial": 17,
    "type": "wall",
    "wIndex": 1,
  },
  {
    "cIndex": 2,
    "initial": 18,
    "type": "wall",
    "wIndex": 2,
  },
  {
    "cIndex": 2,
    "initial": 19,
    "type": "wall",
    "wIndex": 3,
  },
  {
    "cIndex": 2,
    "initial": 20,
    "type": "wall",
    "wIndex": 4,
  },
  {
    "cIndex": 2,
    "initial": 21,
    "type": "wall",
    "wIndex": 5,
  },
  {
    "cIndex": 2,
    "initial": 22,
    "type": "wall",
    "wIndex": 6,
  },
  {
    "cIndex": 2,
    "initial": 23,
    "type": "wall",
    "wIndex": 7,
  },
  {
    "cIndex": 3,
    "initial": 24,
    "type": "wall",
    "wIndex": 0,
  },
  {
    "cIndex": 3,
    "initial": 25,
    "type": "wall",
    "wIndex": 1,
  },
  {
    "cIndex": 3,
    "initial": 26,
    "type": "wall",
    "wIndex": 2,
  },
  {
    "cIndex": 3,
    "initial": 27,
    "type": "wall",
    "wIndex": 3,
  },
  {
    "cIndex": 3,
    "initial": 28,
    "type": "wall",
    "wIndex": 4,
  },
  {
    "cIndex": 3,
    "initial": 29,
    "type": "wall",
    "wIndex": 5,
  },
  {
    "initial": 30,
    "sIndex": 1,
    "type": "special",
  },
  {
    "initial": 31,
    "sIndex": 0,
    "type": "special",
  },
  {
    "initial": 32,
    "sIndex": 2,
    "type": "special",
  },
  {
    "cIndex": 3,
    "initial": 33,
    "type": "wall",
    "wIndex": 6,
  },
  {
    "cIndex": 3,
    "initial": 34,
    "type": "wall",
    "wIndex": 7,
  },
  {
    "cIndex": 4,
    "initial": 35,
    "type": "wall",
    "wIndex": 0,
  },
  {
    "cIndex": 4,
    "initial": 36,
    "type": "wall",
    "wIndex": 1,
  },
  {
    "cIndex": 4,
    "initial": 37,
    "type": "wall",
    "wIndex": 2,
  },
  {
    "cIndex": 4,
    "initial": 38,
    "type": "wall",
    "wIndex": 3,
  },
  {
    "cIndex": 4,
    "initial": 39,
    "type": "wall",
    "wIndex": 4,
  },
  {
    "cIndex": 4,
    "initial": 40,
    "type": "wall",
    "wIndex": 5,
  },
  {
    "cIndex": 4,
    "initial": 41,
    "type": "wall",
    "wIndex": 6,
  },
  {
    "cIndex": 4,
    "initial": 42,
    "type": "wall",
    "wIndex": 7,
  },
  {
    "cIndex": 5,
    "initial": 43,
    "type": "wall",
    "wIndex": 0,
  },
  {
    "cIndex": 5,
    "initial": 44,
    "type": "wall",
    "wIndex": 1,
  },
  {
    "cIndex": 5,
    "initial": 45,
    "type": "wall",
    "wIndex": 2,
  },
  {
    "cIndex": 5,
    "initial": 46,
    "type": "wall",
    "wIndex": 3,
  },
  {
    "cIndex": 5,
    "initial": 47,
    "type": "wall",
    "wIndex": 4,
  },
  {
    "cIndex": 5,
    "initial": 48,
    "type": "wall",
    "wIndex": 5,
  },
  {
    "cIndex": 5,
    "initial": 49,
    "type": "wall",
    "wIndex": 6,
  },
  {
    "cIndex": 5,
    "initial": 50,
    "type": "wall",
    "wIndex": 7,
  },
  {
    "cIndex": 6,
    "initial": 51,
    "type": "wall",
    "wIndex": 0,
  },
  {
    "cIndex": 6,
    "initial": 52,
    "type": "wall",
    "wIndex": 1,
  },
  {
    "cIndex": 6,
    "initial": 53,
    "type": "wall",
    "wIndex": 2,
  },
  {
    "cIndex": 6,
    "initial": 54,
    "type": "wall",
    "wIndex": 3,
  },
  {
    "cIndex": 6,
    "initial": 55,
    "type": "wall",
    "wIndex": 4,
  },
  {
    "cIndex": 6,
    "initial": 56,
    "type": "wall",
    "wIndex": 5,
  },
  {
    "cIndex": 6,
    "initial": 57,
    "type": "wall",
    "wIndex": 6,
  },
  {
    "cIndex": 6,
    "initial": 58,
    "type": "wall",
    "wIndex": 7,
  },
  {
    "cIndex": 7,
    "initial": 59,
    "type": "wall",
    "wIndex": 0,
  },
  {
    "cIndex": 7,
    "initial": 60,
    "type": "wall",
    "wIndex": 1,
  },
  {
    "cIndex": 7,
    "initial": 61,
    "type": "wall",
    "wIndex": 2,
  },
  {
    "cIndex": 7,
    "initial": 62,
    "type": "wall",
    "wIndex": 3,
  },
]
`;
